File: Chapter 10/C#/Section_10_3_Resilience_Decorators/Fragile/FragilePaymentService.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;

namespace Chapter10.Fragile;

/// <summary>
/// THE FRAGILE IMPLEMENTATION:
/// 
/// DESIGN NOTE:
/// This class represents the "Naive" way of handling external dependencies in C#. 
/// It treats a remote network call as if it were a reliable, local constant.
///
/// ARCHITECTURAL CRITIQUE:
/// 1. TEMPORAL COUPLING: This method is "Locked" to the network. Without a timeout 
///    or retry policy, the calling thread is held hostage by FlakyPayments's server response time.
/// 
/// 2. ABSTRACTION LEAK: There is no Interface (Port). The business logic is forced 
///    to depend directly on this concrete implementation and the 'HttpClient' library, 
///    violating the Downward Dependency Rule (Chapter 4).
/// 
/// 3. THE HAPPY PATH FALLACY: The code assumes EnsureSuccessStatusCode() is a 
///    finality. In a distributed system, a 503 (Service Unavailable) is often a 
///    temporary "hiccup" that could be solved by a retry. Here, it is a fatal crash.
/// </summary>
public class FragilePaymentService
{
    private static readonly HttpClient _httpClient = new HttpClient();

    public async Task<string> ChargeCreditCardAsync(decimal amount)
    {
        // This raw function is an exposed liability. 
        // It assumes the network is reliable (The Happy Path Fallacy).
        Console.WriteLine($"      [Payment Service] Attempting to charge ${amount}...");

        // If FlakyPayments is down, this line throws a ConnectionError or Timeout.
        // Since there is no retry logic, the user gets an error and the 
        // business loses money.
        var response = await _httpClient.PostAsJsonAsync(
            "https://api.flakypayments.com/charge", 
            new { amount = amount, order_id = "12345" }
        );

        // This throws an exception if the status code is not 2xx.
        // Failure is absolute.
        response.EnsureSuccessStatusCode();

        return await response.Content.ReadAsStringAsync();
    }
}

File: Chapter 10/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter10/section_10_3_resilience_decorators/fragile/FragilePaymentService.java
package com.grokkingsoftwarearchitecture.chapter10.section_10_3_resilience_decorators.fragile;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;

/**
 * THE FRAGILE IMPLEMENTATION:
 * * DESIGN NOTE:
 * This class represents the "Naive" way of handling external dependencies in Java. 
 * It treats a remote network call as if it were a local, reliable method call.
 *
 * ARCHITECTURAL CRITIQUE:
 * 1. TEMPORAL COUPLING: This method is "Locked" to the network. Without a timeout 
 * or retry policy, the calling thread is held hostage by FlakyPayments's server. If the
 * network hangs, this thread is essentially dead until the OS kills it.
 * * 2. ABSTRACTION LEAK: There is no Interface (Port). The business logic is forced 
 * to depend directly on this concrete implementation and the 'java.net.http' 
 * library, violating the Downward Dependency Rule (Chapter 4).
 * * 3. THE HAPPY PATH FALLACY: The code assumes that if it sends a request, it 
 * receives a success. It provides no mechanism to survive a 503 (Service 
 * Unavailable) or a temporary network blip. Failure here is absolute and total.
 */
public class FragilePaymentService {

    // A raw, unprotected HTTP client.
    private static final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(2))
            .build();

    /**
     * A raw function sitting right in the middle of your business logic.
     * It assumes the outside world is stable and friendly.
     */
    public String chargeCreditCard(double amount) throws Exception {
        System.out.println("      [Payment Service] Attempting to charge $" + amount + "...");

        // Constructing a hardcoded JSON payload
        String jsonBody = "{\"amount\": " + amount + ", \"order_id\": \"12345\"}";

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api.flakypayments.com/charge"))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .build();

        // If FlakyPayments is down or the internet blips, this throws an IOException 
        // and crashes the application flow instantly.
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        // A naive check: if it's not 200, we just throw an exception.
        // No retries, no backoff, no mercy.
        if (response.statusCode() >= 400) {
            throw new RuntimeException("API Error: Status Code " + response.statusCode());
        }

        return response.body();
    }
}